src/collapse/dtrace.rs

```diff
@@ -109,21 +151,152 @@ impl Collapse for Folder {
+
+        crossbeam::thread::scope(|scope| {
+            // Spin up the threadpool / worker threads.
+            let (tx_input, rx_input) = channel::unbounded();
+            let (tx_output, rx_output) = channel::unbounded();
+            let mut handles = Vec::with_capacity(self.opt.nthreads);
+            for _ in 0..self.opt.nthreads {
+                let rx_input = rx_input.clone();
+                let tx_output = tx_output.clone();
+
+                let occurrences = self.occurrences.clone();
+                let opt = self.opt.clone();
+
+                let handle = scope.spawn(move |_| {
+                    let mut folder = Folder {
+                        cache_inlines: Vec::default(),
+                        occurrences,
+                        stack: VecDeque::default(),
+                        stack_str_size: 0,
+                        opt,
+                    };
+                    loop {
+                        // Receive the data...
+                        let data = match rx_input.recv().unwrap() {
+                            Message::Job(data) => data,
+                            Message::Shutdown => return,
+                        };
+                        // Process the data...
+                        let result = folder.collapse_single_threaded(&data[..]);
+                        // Send the result...
+                        tx_output.send(result).unwrap();
+                        // Prepare for the next round...
+                        folder.stack.clear();
+                        folder.stack_str_size = 0;
+                    }
+                });
+                handles.push(handle);
+            }
+
+            // State for the loop...
+            let buf_capacity = collapse::smallest_power_of_two_larger_than(
+                collapse::GUESS_NBYTES_PER_STACK * self.opt.nstacks_per_job,
+            );
+            let mut buf = Vec::with_capacity(buf_capacity);
+            let (mut index, mut njobs, mut nstacks) = (0, 0, 0);
+
+            // The loop...
+            loop {
+                // First, read some data into the `utils::Buffer`.
+                let n = reader.read_until(b'\n', &mut buf)?;
+
+                // If we're at the end of the data...
+                if n == 0 {
+                    // Send the last slice.
+                    tx_input.send(Message::Job(buf)).unwrap();
+                    njobs += 1;
+                    // Exit the loop.
+                    break;
+                }
+
+                let line = &buf[index..index + n];
+                index += n;
+
+                if is_end_of_stack(line) {
+                    // It's the end of a stack; count it
+                    nstacks += 1;
+                    // If we've seen enough stacks to make up a slice...
+                    if nstacks == self.opt.nstacks_per_job {
+                        // Send it.
+                        let chunk = mem::replace(&mut buf, Vec::with_capacity(buf_capacity));
```

Comment:
I _think_ what I'd do here is replace with a `Vec::with_capacity(buf.capacity())`. This would be much better in cases where our initial guess for `buf_capacity` is too low.

Reply:
Great idea.